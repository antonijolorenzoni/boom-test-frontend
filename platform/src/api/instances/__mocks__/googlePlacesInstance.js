export const onFetchGooglePlacesOptions = (value, token) =>
  Promise.resolve([
    {
      value:
        'EjdDb3JzbyBNYWdlbnRhLCBNaWxhbiwgTWV0cm9wb2xpdGFuIENpdHkgb2YgTWlsYW4sIEl0YWx5Ii4qLAoUChIJLfawmlDBhkcRX8dY63xh_1wSFAoSCed1Ej9JwYZHEY0OdMYTzf88',
      label: 'Corso Magenta',
    },
  ]);

export const fetchGoogleAddressDetails = () =>
  Promise.resolve({
    placeId: 'ChIJ4edudVzBhkcRHP7WsU-4aSM',
    street: 'Corso Magenta',
    timezone: 'Europe/Rome',
  });
