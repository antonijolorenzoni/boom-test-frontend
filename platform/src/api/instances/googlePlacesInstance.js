//
// ──────────────────────────────────────────────────────────────────────────────────────────────────────────── I ──────────
//   :::::: A X I O S   I N S T A N C E   F O R   G O O G L E   M A P S   A P I : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
//

import _ from 'lodash';
import moment from 'moment';

const axios = require('axios');

const GOOGLE_API_KEY = process.env.REACT_APP_GOOGLE_API_KEY;

export function geocodeByAddress(address) {
  const geocoder = new window.google.maps.Geocoder();
  const { OK } = window.google.maps.GeocoderStatus;
  return new Promise((resolve, reject) => {
    geocoder.geocode({ address }, (results, status) => {
      if (status !== OK) {
        reject(status);
      }
      resolve(results);
    });
  });
}

export function getLatLng(result) {
  return new Promise((resolve, reject) => {
    try {
      const latLng = {
        latitude: result.geometry.location,
        longitude: result.geometry.location,
      };
      resolve(latLng);
    } catch (e) {
      reject(e);
    }
  });
}

export function geocodeByPlaceId(placeId) {
  const geocoder = new window.google.maps.Geocoder();
  const { OK } = window.google.maps.GeocoderStatus;
  return new Promise((resolve, reject) => {
    geocoder.geocode({ placeId }, (results, status) => {
      if (status !== OK) {
        reject(status);
      }
      resolve(results);
    });
  });
}

export function fetchPlaceFromGoogle(text, token, country) {
  const autocompleteService = new window.google.maps.places.AutocompleteService();
  return new Promise((resolve, reject) => {
    autocompleteService.getPlacePredictions(
      {
        input: text,
        sessiontoken: token,
        componentRestrictions: {
          country: country || null,
        },
        types: ['address'],
      },
      (place, status) => {
        resolve(place);
      }
    );
  });
}

export async function onFetchGooglePlacesOptions(value, token, country) {
  const addresses = await fetchPlaceFromGoogle(value, token, country);
  const options = _.map(addresses, (address) => ({
    value: address.place_id, // workaround for creatable component
    label: address.description,
  }));
  return options;
}

export async function getPlaceTimezone(lat, lon) {
  try {
    const { OK } = window.google.maps.GeocoderStatus;
    const timezoneResponse = await axios.get(
      `https://maps.googleapis.com/maps/api/timezone/json?location=${lat},${lon}&timestamp=${moment().unix()}&key=${GOOGLE_API_KEY}`
    );
    if (timezoneResponse && timezoneResponse.data && timezoneResponse.data.timeZoneId && timezoneResponse.data.status === OK) {
      return timezoneResponse.data.timeZoneId;
    }
    return 'Europe/Rome';
  } catch (error) {
    return 'Europe/Rome';
  }
}

const findInAddressFn = (addressSelected) => (field) =>
  _.find(addressSelected.address_components, (component) => _.indexOf(component.types, field) !== -1);

export const buildAddressDTO = (selectedAddress, timezone) => {
  const location = {
    latitude: selectedAddress.geometry.location.lat,
    longitude: selectedAddress.geometry.location.lng,
  };

  const findFieldInAddress = findInAddressFn(selectedAddress);

  const city = findFieldInAddress('locality');
  const postalCode = findFieldInAddress('postal_code');
  const country = findFieldInAddress('country');
  const street = findFieldInAddress('route');
  const street_number = findFieldInAddress('street_number');

  const addressDTO = {
    location,
    city: city?.long_name,
    street: street?.long_name,
    street_number: street_number?.long_name,
    postalCode: postalCode?.long_name,
    country: country?.long_name,
    countryCode: country?.short_name,
    formattedAddress: selectedAddress.formatted_address,
    placeId: selectedAddress.place_id,
    timezone,
  };

  return addressDTO;
};

export async function fetchGoogleAddressDetails(address, sessionToken) {
  try {
    // need to check if it is an array
    // because react-select trigger the onChange function also when press backspace and the field is empty
    if (!address || Array.isArray(address)) {
      return null;
    }

    const details = await axios.get(
      `https://maps.googleapis.com/maps/api/geocode/json?address=${address.label}&key=${GOOGLE_API_KEY}&sensor=false&language=en&sessiontoken=${sessionToken}`
    );

    const detailsResult = _.get(details, 'data.results');

    if (detailsResult && _.isArray(details.data.results)) {
      const selectedAddress = _.first(details.data.results);

      if (!selectedAddress) {
        return null;
      }

      const latitude = selectedAddress.geometry.location.lat;
      const longitude = selectedAddress.geometry.location.lng;
      const timezone = await getPlaceTimezone(latitude, longitude);

      return buildAddressDTO(selectedAddress, timezone);
    }

    return null;
  } catch (error) {
    throw error;
  }
}
