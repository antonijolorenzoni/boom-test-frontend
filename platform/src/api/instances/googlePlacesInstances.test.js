import { buildAddressDTO } from './googlePlacesInstance';

const fullAddress = {
  address_components: [
    {
      long_name: '85',
      short_name: '85',
      types: ['street_number'],
    },
    {
      long_name: 'Corso Magenta',
      short_name: 'Corso Magenta',
      types: ['route'],
    },
    {
      long_name: 'Milano',
      short_name: 'Milano',
      types: ['locality', 'political'],
    },
    {
      long_name: 'Milano',
      short_name: 'Milano',
      types: ['administrative_area_level_3', 'political'],
    },
    {
      long_name: 'Città Metropolitana di Milano',
      short_name: 'MI',
      types: ['administrative_area_level_2', 'political'],
    },
    {
      long_name: 'Lombardia',
      short_name: 'Lombardia',
      types: ['administrative_area_level_1', 'political'],
    },
    {
      long_name: 'Italy',
      short_name: 'IT',
      types: ['country', 'political'],
    },
    {
      long_name: '20123',
      short_name: '20123',
      types: ['postal_code'],
    },
  ],
  formatted_address: 'Corso Magenta, 85, 20123 Milano MI, Italy',
  geometry: {
    location: {
      lat: 45.4658235,
      lng: 9.1666898,
    },
    location_type: 'ROOFTOP',
    viewport: {
      northeast: {
        lat: 45.4671724802915,
        lng: 9.168038780291502,
      },
      southwest: {
        lat: 45.4644745197085,
        lng: 9.165340819708499,
      },
    },
  },
  place_id: 'ChIJ4edudVzBhkcRHP7WsU-4aSM',
  plus_code: {
    compound_code: 'F588+8M Milan, Metropolitan City of Milan, Italy',
    global_code: '8FQFF588+8M',
  },
  types: ['street_address'],
};

const withoutStreetNumberAddress = {
  address_components: [
    {
      long_name: 'Corso Magenta',
      short_name: 'Corso Magenta',
      types: ['route'],
    },
    {
      long_name: 'Milano',
      short_name: 'Milano',
      types: ['locality', 'political'],
    },
    {
      long_name: 'Milano',
      short_name: 'Milano',
      types: ['administrative_area_level_3', 'political'],
    },
    {
      long_name: 'Città Metropolitana di Milano',
      short_name: 'MI',
      types: ['administrative_area_level_2', 'political'],
    },
    {
      long_name: 'Lombardia',
      short_name: 'Lombardia',
      types: ['administrative_area_level_1', 'political'],
    },
    {
      long_name: 'Italy',
      short_name: 'IT',
      types: ['country', 'political'],
    },
  ],
  formatted_address: 'Corso Magenta, Milano MI, Italy',
  geometry: {
    bounds: {
      northeast: {
        lat: 45.4668262,
        lng: 9.180570699999999,
      },
      southwest: {
        lat: 45.4649824,
        lng: 9.165971299999999,
      },
    },
    location: {
      lat: 45.4657262,
      lng: 9.173159199999999,
    },
    location_type: 'GEOMETRIC_CENTER',
    viewport: {
      northeast: {
        lat: 45.4672532802915,
        lng: 9.180570699999999,
      },
      southwest: {
        lat: 45.4645553197085,
        lng: 9.165971299999999,
      },
    },
  },
  place_id: 'ChIJLfawmlDBhkcRX8dY63xh_1w',
  types: ['route'],
};

const onlyCityAddress = {
  address_components: [
    {
      long_name: 'Milan',
      short_name: 'Milan',
      types: ['locality', 'political'],
    },
    {
      long_name: 'Milan',
      short_name: 'Milan',
      types: ['administrative_area_level_3', 'political'],
    },
    {
      long_name: 'Metropolitan City of Milan',
      short_name: 'MI',
      types: ['administrative_area_level_2', 'political'],
    },
    {
      long_name: 'Lombardy',
      short_name: 'Lombardy',
      types: ['administrative_area_level_1', 'political'],
    },
    {
      long_name: 'Italy',
      short_name: 'IT',
      types: ['country', 'political'],
    },
  ],
  formatted_address: 'Milan, Metropolitan City of Milan, Italy',
  geometry: {
    bounds: {
      northeast: {
        lat: 45.535689,
        lng: 9.2903463,
      },
      southwest: {
        lat: 45.3897787,
        lng: 9.065118199999999,
      },
    },
    location: {
      lat: 45.4642035,
      lng: 9.189982,
    },
    location_type: 'APPROXIMATE',
    viewport: {
      northeast: {
        lat: 45.535689,
        lng: 9.2903463,
      },
      southwest: {
        lat: 45.3897787,
        lng: 9.065118199999999,
      },
    },
  },
  place_id: 'ChIJ53USP0nBhkcRjQ50xhPN_zw',
  types: ['locality', 'political'],
};

const onlyCountryAddress = {
  address_components: [
    {
      long_name: 'Italy',
      short_name: 'IT',
      types: ['country', 'political'],
    },
  ],
  formatted_address: 'Italy',
  geometry: {
    bounds: {
      northeast: {
        lat: 47.092,
        lng: 18.7975999,
      },
      southwest: {
        lat: 35.4897,
        lng: 6.6267201,
      },
    },
    location: {
      lat: 41.87194,
      lng: 12.56738,
    },
    location_type: 'APPROXIMATE',
    viewport: {
      northeast: {
        lat: 47.092,
        lng: 18.7975999,
      },
      southwest: {
        lat: 35.4897,
        lng: 6.6267201,
      },
    },
  },
  place_id: 'ChIJA9KNRIL-1BIRb15jJFz1LOI',
  types: ['country', 'political'],
};

describe('googlePlaceInstances', () => {
  describe('buildAddressDTO', () => {
    it('build a full address', () => {
      expect(buildAddressDTO(fullAddress, 'Europe/Rome')).toMatchSnapshot();
    });

    it('build an address without street number', () => {
      expect(buildAddressDTO(withoutStreetNumberAddress, 'Europe/Rome')).toMatchSnapshot();
    });

    it('build an address composed by only a city', () => {
      expect(buildAddressDTO(onlyCityAddress, 'Europe/Rome')).toMatchSnapshot();
    });

    it('build an address composed by only a country', () => {
      expect(buildAddressDTO(onlyCountryAddress, 'Europe/Rome')).toMatchSnapshot();
    });
  });
});
