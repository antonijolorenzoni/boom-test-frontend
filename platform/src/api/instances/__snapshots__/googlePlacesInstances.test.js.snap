// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`googlePlaceInstances buildAddressDTO build a full address 1`] = `
Object {
  "city": "Milano",
  "country": "Italy",
  "countryCode": "IT",
  "formattedAddress": "Corso Magenta, 85, 20123 Milano MI, Italy",
  "location": Object {
    "latitude": 45.4658235,
    "longitude": 9.1666898,
  },
  "placeId": "ChIJ4edudVzBhkcRHP7WsU-4aSM",
  "postalCode": "20123",
  "street": "Corso Magenta",
  "street_number": "85",
  "timezone": "Europe/Rome",
}
`;

exports[`googlePlaceInstances buildAddressDTO build an address composed by only a city 1`] = `
Object {
  "city": "Milan",
  "country": "Italy",
  "countryCode": "IT",
  "formattedAddress": "Milan, Metropolitan City of Milan, Italy",
  "location": Object {
    "latitude": 45.4642035,
    "longitude": 9.189982,
  },
  "placeId": "ChIJ53USP0nBhkcRjQ50xhPN_zw",
  "postalCode": undefined,
  "street": undefined,
  "street_number": undefined,
  "timezone": "Europe/Rome",
}
`;

exports[`googlePlaceInstances buildAddressDTO build an address composed by only a country 1`] = `
Object {
  "city": undefined,
  "country": "Italy",
  "countryCode": "IT",
  "formattedAddress": "Italy",
  "location": Object {
    "latitude": 41.87194,
    "longitude": 12.56738,
  },
  "placeId": "ChIJA9KNRIL-1BIRb15jJFz1LOI",
  "postalCode": undefined,
  "street": undefined,
  "street_number": undefined,
  "timezone": "Europe/Rome",
}
`;

exports[`googlePlaceInstances buildAddressDTO build an address without street number 1`] = `
Object {
  "city": "Milano",
  "country": "Italy",
  "countryCode": "IT",
  "formattedAddress": "Corso Magenta, Milano MI, Italy",
  "location": Object {
    "latitude": 45.4657262,
    "longitude": 9.173159199999999,
  },
  "placeId": "ChIJLfawmlDBhkcRX8dY63xh_1w",
  "postalCode": undefined,
  "street": "Corso Magenta",
  "street_number": undefined,
  "timezone": "Europe/Rome",
}
`;
